### ISSUE TITLE
Retry logic in clients

### ISSUE THREAD

**Issue by cmccoy** [brianfrankcooper/YCSB#334]
Right now the Cassandra Thrift clients (0.7, 0.8, and 1.0) and the Aerospike client (#324) implement client-side operation retries within YCSB code. I find that a little counterintuitive: I'd expect the operation failure rates reported by YCSB to be based on pretty simplistic client usage. Anyone have any thoughts?

**Comment by busbey** [comment]
I can see the merit in both arguments, but having differences across stores is yet more for our users to grok while comparing. So I think this is something we need to set a guideline for and document (sort of like the issue of making sure clients flush to remote after a transaction).

If we take the position of no-retries, what do we do with a datastore binding that does the retry below us? I guess at that point it can be seen as a feature of using that datastore?

If we want retries in general, it seems better to pull that functionality up into the wrapper. then we can report on the error rate still.

**Comment by cmccoy** [comment]
> If we take the position of no-retries, what do we do with a datastore binding that does the retry below us? I guess at that point it can be seen as a feature of using that datastore?

I think so. If the client itself implements retries, fine - you'd see the same behavior in the wild.

**Comment by busbey** [comment]
yeah, I agree. We should start labeling these issues that have datastore binding guidelines. Eventually once we make a doc we can then do a sweep of bringing bindings in line. Not much use enforcing it until then.

**Comment by tlopatic** [comment]
Agreed, valid point.

I added the retry logic to take care of the occasional Amazon EBS hiccup. Thinking about this again, though, that logic should actually go into the Aerospike client and not into the YCSB binding, so that everyone benefits from it.

I'll change that.
